### PR TITLE
tcp client: make close type configurable

### PR DESCRIPTION
**Pull request by botengyao**
make async tcp client close type configurable.

Commit Message:
Additional Description:
Risk Level: low
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: envoy/tcp/async_tcp_client.h [hunk 1]
```diff
@@ -50,6 +50,12 @@ class AsyncTcpClient {
    */
   virtual void close() PURE;
 
+  /**
+   * Close the client. It closes the connection based on close type.
+   * @param type the connection close type.
+   */
+  virtual void close(Network::ConnectionCloseType type) PURE;
```
Comment (yanavlasov) [inline envoy/tcp/async_tcp_client.h:57]: Given that this is new API and there is no need for backward compatibility, I would just go with the one `close` method that requires the close type parameter.

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

File: envoy/tcp/async_tcp_client.h [hunk 2]
```diff
@@ -45,10 +45,11 @@ class AsyncTcpClient {
   virtual bool connect() PURE;
 
   /**
-   * Close the client. It abortively closes the connection discarding any unsent data.
+   * Close the client. It closes the connection based on close type.
    * The underlying connection will be defer deleted when a Close event is received.
+   * @param type the connection close type.
    */
-  virtual void close() PURE;
+  virtual void close(Network::ConnectionCloseType type) PURE;
```
Comment (kyessenov) [inline envoy/tcp/async_tcp_client.h:52]: What's the recommended value here for the parameter?

**Comment by alyssawilk** [comment]
@yanavlasov ping?

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**